src/utils/csrf-utils.ts

```diff
@@ -0,0 +1,98 @@
+'use server';
+
+import { jwtDecode } from 'jwt-decode';
+import { createHmac, randomBytes } from 'node:crypto';
+import { cookies } from 'next/headers';
+import { getEnvironmentVariable } from './get-environment-variable';
+import { getAccessTokenServer } from './amplify-utils';
+
+export const getCsrfFormValue = async () =>
+  cookies().get('csrf_token')?.value ?? 'no_token';
+
+export const getSessionId = async () => {
+  const accessToken = await getAccessTokenServer();
+
+  if (!accessToken) {
+    throw new Error('Could not get access token');
+  }
+
+  const jwt = jwtDecode(accessToken);
+
+  const sessionId = jwt.jti;
+
+  if (!sessionId) {
+    throw new Error('Could not get session ID');
+  }
+
+  return sessionId;
+};
+
+export const generateCsrf = async () => {
+  const secret = getEnvironmentVariable('CSRF_SECRET');
+  const sessionId = await getSessionId();
+
+  const salt = randomBytes(8).toString('hex');
+
+  const hash = createHmac('sha256', secret)
+    .update(sessionId)
+    .update(salt)
+    .digest('hex');
+
+  const csrfToken = `${hash}.${salt}`;
+
+  cookies().set('csrf_token', csrfToken, {
+    httpOnly: true,
+    secure: true,
+  });
+
+  return csrfToken;
+};
+
+export const verifyCsrfToken = async (
+  csrfToken: string,
+  secret: string,
+  sessionId: string
+) => {
+  const [hmac, salt] = csrfToken.split('.');
```

Comment:
Just been thinking about this. Does it need validating here that the token has the right format with the dot? If `salt` is undefined because of the format, then I think you will get an error thrown here when you pass it to hmac.update